binutils/strip.1
.\" Copyright (c) 1991 Free Software Foundation
.\" See section COPYING for conditions for redistribution
.TH strip 1 "5 November 1991" "cygnus support" "GNU Development Tools"
.de BP
.sp
.ti \-.2i
\(**
..

.SH NAME
strip\(em\&Discard symbols from object files.

.SH SYNOPSIS
.hy 0
.na
.TP
.B strip
.RB "[\|" \-s | +strip-all "\|]" 
.RB "[\|" \-g | \-S | +strip-debug\c
\|]
.RB "[\|" \-x | +discard-all "\|]" 
.RB "[\|" \-X | +discard-locals "\|]"
.RB "[\|" "\-T\ "\c
.I bfdname\c
\&\|]
.I objfiles\c
\&.\|.\|.
.ad b
.hy 1
.SH DESCRIPTION
GNU \c
.B strip\c
\& will discard all symbols from object files
\c
.I objfiles\c
\&, if no options are specified; or only certain symbols,
depending on its command-line options.

\c
.B strip\c
\& will not execute unless at least one object file is listed.

.I WARNING:\c
\& \c
.B strip\c
\& modifies the files named in its argument,
rather than writing modified copies under different names.

.SH OPTIONS
The long and short forms of options, shown here as alternatives, are
equivalent.

.TP
.B \-s
.TP
.B +strip-all 
This is the default case: strip all symbol entries from \c
.I objfiles\c
\&.

.TP
.B \-g
.TP
.B \-S
.TP
.B +strip-debug 
Discard only debugging symbol information from \c
.I objfiles\c
\&.

.TP
.B \-x
.TP
.B +discard-all 
Discard all symbols local to each file in \c
.I objfiles\c
\&.
\c
.I WARNING:\c
\& Note that \c
.B +discard-all\c
\& discards only \c
.I local\c
\&
symbols, in spite of its name.

.TP
.B \-X
.TP
.B +discard-locals 
Discard local symbols starting with `\|\c
.B L\c
\|' from each file in
\c
.I objfiles\c
\&.  (Some compilers produce internally-used symbols that
begin with `\|\c
.B L\c
\|'.)

.TP
.BI "-T " "bfdname"\c
\&
You can specify a particular object-code format \c
.I bfdname\c
\& for
\c
.I objfiles\c
\&.  This may not be necessary; \c
.I strip\c
\& can automatically
recognize many formats.   for information on listing
available formats.
.PP

.SH "SEE ALSO"
.RB "`\|" binutils "\|'" 
entry in 
.B
info\c
\&; 
.I
The GNU Binary Utilities\c
\&, Roland H. Pesch (October 1991).

.SH COPYING
Copyright (c) 1991 Free Software Foundation, Inc.
.PP
Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the
entire resulting derived work is distributed under the terms of a
permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this
manual into another language, under the above conditions for modified
versions, except that this permission notice may be included in
translations approved by the Free Software Foundation instead of in
the original English.


ld/NEWS
-*- text -*-

* All SFrame sections generated by the linker have the header flag
  SFRAME_F_FDE_FUNC_START_ADDR_PCREL set to indicate the new encoding for
  sfde_func_start_address field in the SFrame section.  Relocatable SFrame
  links are now fixed.

* On RISC-V, add new PLT formats, and GNU property merge rules for zicfiss and
  zicfilp extensions.

* The linker's --stats option can take an optional argument which if used is
  interpreted as a filename into which resource usage information should be
  stored.  As an alternative mechanism the LD_STATS environment variable can
  also be used to achieve the same results.  Resource usage information for
  various phases of the linking operation is now included in the report.
  If a map file is being produced then the information is also included there.
  The --no-stats option can be used to disable stat reporting, should it have
  been enabled.

* Remove the linker -taso option for Alpha target, as Linux/Alpha kernel
  support for 32-bit pointers has been removed.

* On avr, the default linker scripts now assert that the .progmem sections
  don't extend past 0xffff since they are accessed by means of LPM.
  For data in program memory that may be located past 0xffff, a .progmemx
  section should be used.

* On s390, generate ".eh_frame" unwind information for the linker generated
  .plt section.  Enabled by default.  Can be disabled using linker option
  --no-ld-generated-unwind-info.

* On s390, add support for linker option --[no-]ld-generated-unwind-info.

Changes in 2.44:

* Support for Nios II target has been removed, as this architecture has been
  EOL'ed by Intel.

* The default maximum page size was changed from 16KiB to 64KiB for LoongArch.

* Add support for mixed LTO and non-LTO codes in relocatable output.

* Add --image-base=<ADDR> option to the ELF linker to behave the same
  as -Ttext-segment for compatibility with LLD.

* Add a "--build-id=xx" option, if built with the xxhash library.  This
  produces a 128-bit hash, 2-4x faster than md5 or sha1.

* The ELF linker option --package-metadata supports percent-encoded and
  %[string] encoded JSON payload now to ease passing around this option
  without getting the JSON payload corrupted.

Changes in 2.43:

* Add support for LoongArch DT_RELR (compressed R_LARCH_RELATIVE).

* Put .got .got.plt in the relro to make it read-only after relocation.

* Add -z isa-level-report=[none|all|needed|used] to the x86 ELF linker
  to report needed and used x86-64 ISA levels.

* Add --rosegment option which changes the -z separate-code option so that
  only one read-only segment is created (instead of two).  (The option name
  is misleading, but it matches the name used by LLD and GOLD).

* Add --section-ordering-file <FILE> option to add extra mapping of input
  sections to output sections.

* Add -plugin-save-temps to store plugin intermediate files permanently.

* Add -z memory-seal/-z nomemory-seal options to ELF linker to mark the
  object to memory sealed.   Also added --enable-memory-seal configure option
  to enable the memory sealing by default.

Changes in 2.42:

* Add -z mark-plt/-z nomark-plt options to x86-64 ELF linker to mark PLT
  entries with DT_X86_64_PLT, DT_X86_64_PLTSZ and DT_X86_64_PLTENT dynamic
  tags.  Also added --enable-mark-plt configure option to mark PLT entries
  by default.

* Support Intel APX relocations.

* On RISC-V, add ld target option --[no-]check-uleb128.  Should rebuild the
  objects by binutils 2.42 and up if enabling the option and get warnings,
  since the non-zero addend of SUB_ULEB128 shouldn't be generated from .uleb128
  directives.

* Add support for the KVX instruction set.

* A new linker script sorting directive has been added: REVERSE.  This reverses
  the order of the sorting.  It may be combined with either SORT_BY_NAME or
  SORT_BY_INIT_PRIORITY.

* Added --warn-execstack-objects to warn about executable stacks only when an
  input object file requests one.  Also added --error-execstack and
  --error-rxw-segments options to convert warnings about executable stacks and
  segments into errors.

  Also added --enable-error-execstack=[yes|no] and
  --enable-error-rwx-segments=[yes|no] configure options to set the default for
  converting warnings into errors.

* On LoongArch, various linker relaxation bugs are fixed; the most notable
  of which is BZ 30944 (incorrect .balign semantics).

* On LoongArch, the LoongArch ABI v2.30 (LoongArch ELF psABI v20231219) is
  now implemented. This includes new relocation types, and changed
  semantics for PC-relative relocations handling the higher half of 64-bit
  offsets.

* On LoongArch, link-time TLS optimization and TLS relaxation are now
  supported.

Changes in 2.41:

* The linker now accepts a command line option of --remap-inputs
  <PATTERN>=<FILE> to relace any input file that matches <PATTERN> with
  <FILE>.  In addition the option --remap-inputs-file=<FILE> can be used to
  specify a file containing any number of these remapping directives.

* The linker command line option --print-map-locals can be used to include
  local symbols in a linker map.  (ELF targets only).

* For most ELF based targets, if the --enable-linker-version option is used
  then the version of the linker will be inserted as a string into the .comment
  section.

* The linker script syntax has a new command for output sections: ASCIZ "string"
  This will insert a zero-terminated string at the current location.

* Add command-line option, -z nosectionheader, to omit ELF section
  header.

Changes in 2.40:

* The linker has a new command line option to suppress the generation of any
  warning or error messages.  This can be useful when there is a need to create
  a known non-working binary.  The option is -w or --no-warnings.

* ld now supports zstd compressed debug sections.  The new option
  --compress-debug-sections=zstd compresses debug sections with zstd.


binutils/strip.c
/*** strip.c -- strip certain symbols from a rel file.
   Copyright (C) 1986, 1990 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 1, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

/* BUGS: When there's not enough memory, this should do the copy
   in pieces rather than just fail as it does now */

#include "sysdep.h"
#include "bfd.h"

#include "getopt.h"



#include <signal.h>

/* Various program options */

int show_version = 0;

/* Which symbols to remove. */
enum strip_action {
  strip_undef,
  strip_all,			/* strip all symbols */
  strip_debug,			/* strip all debugger symbols */
} strip_symbols;

/* Which local symbols to remove. */
enum {
  locals_undef,
  locals_start_L,		/* discard locals starting with L */
  locals_all,			/* discard all locals */
} discard_locals;

extern char *mktemp();

/* IMPORTS */
extern char *program_version;
extern char *program_name;
extern char *target;
extern char *xmalloc();

PROTO(static boolean, strip_file, (char *filetostrip));
PROTO(static void, copy_sections, (bfd *ibfd, sec_ptr isection, bfd *obfd));
PROTO(static void, setup_sections, (bfd *ibfd, sec_ptr isection, bfd *obfd));

/** main, etc */

static void
usage ()
{
  fprintf (stderr, "strip %s\nUsage: %s [-gsxSX] files ...\n",
	   program_version, program_name);
  exit (1);
}

struct option long_options[] = {{"strip-all", 0, 0, 's'},
				{"strip-debug", 0, 0, 'S'},
				{"discard-all", 0, 0, 'x'},
				{"discard-locals", 0, 0, 'X'},
				{0, 0, 0, 0}
				};

int
main (argc, argv)
     char **argv;
     int argc;
{
  int ind;
  int c;
  program_name = argv[0];
	
  strip_symbols = strip_undef;	/* default is to strip everything.  */
  discard_locals = locals_undef;
	
  while ((c = getopt_long (argc, argv, "gsST:xX", long_options, &ind)) != EOF) {
    switch (c) {
    case 0:
      break;
    case 's':
      strip_symbols = strip_all;
      break;
    case 'g':
    case 'S':
      strip_symbols = strip_debug;
      break;
    case 'T':
      target = optarg;
      break;
    case 'x':
      discard_locals = locals_all;
      break;
    case 'X':
      discard_locals = locals_start_L;
      break;
    default:
      usage ();
    }
  }

  /* Default is to strip all symbols: */
  if (strip_symbols == strip_undef && discard_locals == locals_undef) {
    strip_symbols = strip_all;
  }

  /* OK, all options now parsed.  If no filename specified, do a.out. */
  if (optind == argc) return !strip_file ("a.out");
  
  /* We were given several filenames to do: */
  while (optind < argc)
    if (!strip_file (argv[optind++])) return 1;

  return 0;
}

/** Hack signals */

/* Why does strip need to do this, and anyway, if it does shouldn't this be
   handled by bfd? */

static int delayed_signal;

static int sigint_handled = 0;
static int sighup_handled = 0;
static int sigterm_handled = 0;

void
delay_signal (signo)
     int signo;
{
  delayed_signal = signo;
  signal (signo, delay_signal);
}

/* Effectively defer handling of asynchronous kill signals.  */
void
handle_sigs ()			/* puff puff */
{
  delayed_signal = 0;
	
  if (signal (SIGINT, SIG_IGN) != SIG_IGN) {
    sigint_handled = 1;
    signal (SIGINT, delay_signal);
  }
	
  if (signal (SIGHUP, SIG_IGN) != SIG_IGN) {
    sighup_handled = 1;
    signal (SIGHUP, delay_signal);
  }
	
  if (signal (SIGTERM, SIG_IGN) != SIG_IGN) {
    sigterm_handled = 1;
    signal (SIGTERM, delay_signal);
  }
	
  return;
}

/* Effectively undefer handling.  */
void
unhandle_sigs ()		/* put them down */
{
  if (sigint_handled) signal (SIGINT, SIG_DFL);

  if (sighup_handled) signal (SIGHUP, SIG_DFL);

  if (sigterm_handled) signal (SIGTERM, SIG_DFL);
	
  /* Handle any signal that came in while they were deferred.  */
  if (delayed_signal)
    kill (getpid (), delayed_signal);
	
  return;
}


static boolean
strip_file (filetostrip)
     char *filetostrip;
{
  static char template[] = "stXXXXXX";
  char *slash;
  char *tmpname;
  bfd *ibfd;
  bfd *obfd;

  ibfd = bfd_openr (filetostrip, target);

  if (ibfd == NULL) bfd_fatal (filetostrip);

  handle_sigs ();		/* light up */

  if (!bfd_check_format (ibfd, bfd_object)) {
    fprintf (stderr, "Can't strip %s file %s.\n",
	     bfd_format_string (bfd_get_format (ibfd)), filetostrip);
    exit (1);
  }

  slash = strrchr( filetostrip, '/' );
  if ( slash ){
    *slash = 0;
    tmpname = xmalloc( strlen(filetostrip) + sizeof(template) + 1 );
    strcpy( tmpname, filetostrip );
    strcat( tmpname, "/" );
    strcat( tmpname, template );
    mktemp( tmpname );
    *slash = '/';
  } else {
    tmpname = xmalloc( sizeof(template) );
    strcpy( tmpname, template );
    mktemp( tmpname );
  }

  obfd = bfd_openw (mktemp(tmpname), (target ? target : bfd_get_target (ibfd)));
  if (obfd == NULL) bfd_fatal (tmpname);

  if (!bfd_set_format (obfd, bfd_get_format (ibfd)))
    bfd_fatal (tmpname);


  if ((bfd_set_start_address (obfd, bfd_get_start_address (ibfd)) == false) ||
      (bfd_set_file_flags (obfd, (bfd_get_file_flags (ibfd) &
				  ~(HAS_LINENO | HAS_DEBUG | HAS_SYMS |
				    HAS_LOCALS))) == false) ||
      bfd_set_start_address (obfd, bfd_get_start_address (ibfd)) == false)
    bfd_fatal (bfd_get_filename (ibfd));

   /* Copy architecture of input file to output file */
   if (!bfd_set_arch_mach (obfd, bfd_get_architecture (ibfd),
 			       bfd_get_machine (ibfd))) {
     fprintf(stderr, "Output file cannot represent architecture %s",
 	bfd_printable_arch_mach (bfd_get_architecture(ibfd),
 				 bfd_get_machine (ibfd)));
   }
 

  /* bfd mandates that all output sections be created and sizes set before
     any output is done.  Thus, we traverse all sections twice.  */
  bfd_map_over_sections (ibfd, setup_sections, (void *)obfd);
  bfd_map_over_sections (ibfd, copy_sections, (void *)obfd);

  if (!bfd_close (obfd)) bfd_fatal (filetostrip);
  if (!bfd_close (ibfd)) bfd_fatal (filetostrip);

  rename(tmpname, filetostrip);
  free(tmpname);

  unhandle_sigs();

  return true;
}

/** Actually do the work */
static void
setup_sections (ibfd, isection, obfd)
     bfd *ibfd;
     sec_ptr isection;
     bfd *obfd;
{
  sec_ptr osection;
  char *err;

  osection = bfd_make_section (obfd, bfd_section_name (ibfd, isection));
  if (osection == NULL) {
      err = "making";
      goto loser;
    }

  if (!bfd_set_section_size(obfd, osection, bfd_section_size(ibfd, isection))) {
     err = "size";
     goto loser;
  }

  if (!bfd_set_section_vma (obfd, osection, bfd_section_vma (ibfd, isection))) {
     err = "vma";
     goto loser;
  }

  if (bfd_set_section_alignment (obfd, osection,
			     bfd_section_alignment (ibfd, isection))
      != true) {
	  err = "alignment";
	  goto loser;
  } /* on error, I presume. */

  if (!bfd_set_section_flags (obfd, osection,
			      bfd_get_section_flags (ibfd, isection))) {
     err = "flags";
     goto loser;
  }

  /* All went well */
  return;
  
 loser:
  fprintf (stderr, "%s: file \"%s\", section \"%s\": error in %s: %s\n",
	   program_name,
	   bfd_get_filename (ibfd), bfd_section_name (ibfd, isection),
	   err, bfd_errmsg (bfd_error));
  exit (1);
}

static void
copy_sections (ibfd, isection, obfd)
     bfd *ibfd;
     sec_ptr isection;
     bfd *obfd;
{
  static unsigned char *memhunk = NULL;
  static unsigned memhunksize = 0;

  sec_ptr osection;
  unsigned long size;
  flagword iflg;
  unsigned char *temp;

  osection = bfd_get_section_by_name (obfd, bfd_section_name (ibfd, isection));

  size = bfd_section_size (ibfd, isection);
  iflg = bfd_get_section_flags (ibfd, isection);

  /* either:
     we don't need any memory because there's nothing in this section,
     we had no memory so we got some,
     we had some memory but not enough so we got more,
     or we fail to allocat. */

  if (size == 0)
     return;

  if ((iflg & SEC_HAS_CONTENTS) == 0)
      return;

  if (memhunk == NULL) {
     memhunk = (unsigned char *) xmalloc (size);
     memhunksize = size;
  }

  if (size > memhunksize) {
     temp = (unsigned char *) xrealloc ((char *) memhunk, size);
     memhunksize = size;
     memhunk = temp;
  }

  /* now we have enough memory, just do it: */
  if (!bfd_get_section_contents (ibfd, isection, memhunk, 0, size))
     bfd_fatal (bfd_get_filename (ibfd));

  if (!bfd_set_section_contents (obfd, osection, memhunk, 0, size))
     bfd_fatal (bfd_get_filename (obfd));
}
